### PR TITLE
Move work items top and bottom for order of execution  (#558)

### PR DESCRIPTION
**Pull request by nimishamukherjee**
This PR will add the UI behaviour only, for moving a work item to either the top or bottom of the visible screen.
UI Task - #558
Note: Current implementation of Move to Bottom will take the user to the end of the loaded list.

### CODE REVIEW

**Comment by joshuawilson** [comment]
@nimishamukherjee will this fix also focus the list on the moved item? @almighty/uxd can you confirm the desired behavior, please?

**Comment by nimishamukherjee** [comment]
@joshuawilson / @almighty/uxd , current behaviour:
when we click on the kebab menu, the work item gets selected and remains selected after being moved to the top or bottom.
1. Select a work item to move
move1

2. Move the work item to top
move2

3. Move the work item to the bottom
move3

**Comment by joshuawilson** [comment]
@almighty/uxd can you confirm that 'move to top' and 'move to bottom' is only moving to top and bottom of the viewable area?

**Comment by mindreeper2420** [comment]
@nimishamukherjee @almighty/uxd If the row is being displayed as selected, then the Detail View should be shown. After the row has been moved, it should fade back to de-selected.

**Comment by joshuawilson** [comment]
@mindreeper2420 what about my question of "what is the bottom"? Bottom of the viewport or bottom of the list?

**Comment by mindreeper2420** [comment]
@joshuawilson the bottom is the bottom of the loaded list.

**Comment by joshuawilson** [comment]
@mindreeper2420 ok, if we have a total list of 100 and we are at the top of the list, we have loaded 20 items. If I tell the item at position 5 to move to the bottom, does it go to 20 or 100? What you said is the loaded list and that is 20. I would expect it to go to 100. 

I have a similar expectation that if I'm on 72 and I tell it to move to the top it should 1.

**Comment by nimishamukherjee** [comment]
@mindreeper2420 - "If the row is being displayed as selected, then the Detail View should be shown. After the row has been moved, it should fade back to de-selected."

Few questions around this:
1. Does selecting the check-box imply selection of work item + opening of detailed view? Would the reverse hold good - deselecting the check-box would deselect the work item and close the detailed view?

2. Does selecting a work item for viewing the details select the check-box and display the top move button?

3. If the user wants to move a work item up by 3 rows, he would have to select the work item and click on "move up" three times. In this scenario, how would the detail view behave, if it was opened initially?

**Comment by mindreeper2420** [comment]
@nimishamukherjee 

1. Correct - selecting the checkbox does not open the detail view.

2. Correct - the move dropdown is not enabled, as nothing has been selected to move. The move functionality in the kabob is neither enabled or disabled, though, so that is just there, but hidden behind the detail panel.

3. If the detail view is open at the time of selection and movement, then it would remain open until closed by the user. Moving items in the list does not force the detail panel to close.

**Comment by nimishamukherjee** [comment]
@mindreeper2420 need your input on @joshuawilson 'squestion:

> @mindreeper2420 ok, if we have a total list of 100 and we are at the top of the list, we have loaded 20 items. If I tell the item at position 5 to move to the bottom, does it go to 20 or 100? What you said is the loaded list and that is 20. I would expect it to go to 100. I have a similar expectation that if I'm on 72 and I tell it to move to the top it should 1.

**Comment by mindreeper2420** [comment]
@nimishamukherjee @joshuawilson a question I have before this is answered: is there a mechanism that informs the user there are more items that what have been loaded? In a standard paginated list, that would be easy with the pagination controls in PF. With the single-page list, there is normally a 20 of 100, etc. Is the UI showing that information?

If yes, then sending the item to the bottom makes sense to go the number 100. If not, then sending it to 100 when only 20 are visible (and perhaps known) would make it confusing and frustrating.

**Comment by joshuawilson** [comment]
@mindreeper2420 there is support for display "20 of 100 items loaded", we just didn't have a requirement nor a design for it yet. I think we should add the "of total" and make move to bottom be the actual end of the list.

**Comment by mindreeper2420** [comment]
@joshuawilson agreed - I think that makes the most sense.

**Comment by Mgranfie** [comment]
This was an an early design and when the implementation went to pagination
and then back to infinite scroll, it go lost in the shuffle.

The number is located here in the Visual Designs.

https://redhat.invisionapp.com/share/NU8YPHLJP#/195933683_A-0100

It needs to be tweaked to fall into place with our current layouts. It
would be great if Brian and Nimishia could connect on this.

Thanks All
-Monica

**Comment by nimishamukherjee** [comment]
[test]

**Comment by vikram-raj** [comment]
@nimishamukherjee When I move WI to top or bottom and then click on that WI then It's open wrong detail page. 

It is working fine for the first time.

**Comment by nimishamukherjee** [comment]
Thanks @vikram-raj good catch

File: src/app/work-item/work-item.service.ts [hunk 1]
```diff
@@ -435,16 +435,24 @@ export class WorkItemService {
    * ToDo: Integrate backend when available, also move by one
    * place should be implemented
    */
-  moveItem(wi: WorkItem, dir: String): void {
+moveItem(wi: WorkItem, dir: String): Promise<WorkItem[]> {
```
Comment (sanbornsen) [inline src/app/work-item/work-item.service.ts]: Indentation missing.

Comment (sanbornsen) [inline src/app/work-item/work-item.service.ts]: Do you need to return WorkItem[] as promise? Can't it just be void?

File: src/app/work-item/work-item.service.ts [hunk 2]
```diff
@@ -480,6 +488,9 @@ export class WorkItemService {
     //console.log(wi.attributes.previousitem, ':' , wi.attributes.nextitem);
     //build the map to reset the indices
     this.buildWorkItemIdIndexMap();
+    return new Promise((resolve, reject) => {
```
Comment (sanbornsen) [inline src/app/work-item/work-item.service.ts]: Is this necessary?

File: src/app/work-item/work-item.service.ts [hunk 3]
```diff
@@ -504,7 +513,7 @@ export class WorkItemService {
         return updatedWorkItem;
       })
       .catch (this.handleError);*/
-  }
+    }
```
Comment (vikram-raj) [inline src/app/work-item/work-item.service.ts]: Extra space.

**Pull request merged**